### PR TITLE
Added Tests to Verify Notifications Are Shown When Notebook Builds Are Not Started(ODS-1347)

### PR DESCRIPTION
**Pull request by prudhvi-165**
Signed-off-by: prudhvi <pnammi@redhat.com>

### CODE REVIEW

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 1]
```diff
@@ -172,3 +191,27 @@ Check GPU Resources
...
+    FOR    ${image}    IN    @{images}
+        RHODS Notification Drawer Should Contain    message=${image}
+    END
+
+Start Remaining Builds And Close Browser
+    [Documentation]     Starts Remaining builds and closes browser
+    ${builds_to_be_created}=  Create List  s2i-pytorch-gpu-cuda-11.4.2-notebook  s2i-tensorflow-gpu-cuda-11.4.2-notebook
+    FOR    ${build_config}    IN    @{builds_to_be_created}
+        ${build_name}=  Start New Build    namespace=redhat-ods-applications    buildconfig=${build_config}
+        Wait Until Build Status Is    namespace=redhat-ods-applications    build_name=${build_name}  expected_status=Complete
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:215]: pytorch and tensorflow can run in parallel so you can avoid to check the build status here, and you could check it once both have started

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 2]
```diff
@@ -115,6 +116,24 @@ Search and Verify GPU Items Appears In Resources Page
     Search Items In Resources Section   GPU
     Check GPU Resources
 
+Verify Notifications Are Shown When Notebook Builds Have Not Started
+    [Documentation]     Verifies that Notifications are shown in RHODS Dashboard when Notebook builds haven't started
+    [Tags]    Sanity
+    ...       ODS-1347
+    ${builds_to_be_deleted}=  Create List  pytorch  tensorflow  minimal  cuda-s2i-thoth
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:123]: could you move the variable definitions (line 123 and 125) to `*** Variables ***` section?

Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Updated.

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 1 above)
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Updated.

**Comment by prudhvi-165** [comment]
Jenkins Link:  https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/747/
Result: Passed

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 3]
```diff
@@ -271,3 +292,27 @@ Filter By Using More Than One Filter And Check Output
     FOR    ${id}    IN    @{LIST_OF_IDS_FOR_COMBINATIONS}
         Deselect Checkbox Using Id    id=${id}
     END
+
+Delete Multiple Builds
```
Comment (jgarciao) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:296]: This could be a keyword in tests/Resources/Page/OCPDashboard/Builds/Builds.robot, having  namespace and build_name_list as parameters

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 4]
```diff
@@ -151,6 +154,24 @@ Verify "Notebook Images Are Building" Is Not Shown When No Images Are Building
     Wait Until All Builds Are Complete  namespace=redhat-ods-applications
     RHODS Notification Drawer Should Not Contain  message=Notebooks images are building
 
+Verify Notifications Are Shown When Notebook Builds Have Not Started
+    [Documentation]     Verifies that Notifications are shown in RHODS Dashboard when Notebook builds haven't started
+    [Tags]    Sanity
```
Comment (jgarciao) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:159]: It should be also in Tier1

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 4 above)
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:158]: Updated.

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 3 above)
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Updated.

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 5]
```diff
@@ -274,3 +295,20 @@ Filter By Using More Than One Filter And Check Output
...
+        RHODS Notification Drawer Should Contain    message=${image}
+    END
+
+Wait Until Remaining Builds Are Complete And Close Browser
+    [Documentation]     Waits Until Remaining builds have Status as Complete and Closes Browser
+    ${pytorch_build}=  Search Last Build    namespace=redhat-ods-applications    build_name_includes=pytorch
+    ${tensorflow_build}=  Search Last Build    namespace=redhat-ods-applications    build_name_includes=tensorflow
+    Wait Until Build Status Is    namespace=redhat-ods-applications    build_name=${pytorch_build}  expected_status=Complete
+    Wait Until Build Status Is    namespace=redhat-ods-applications    build_name=${tensorflow_build}  expected_status=Complete
+    Dashboard Test Teardown
```
Comment (bdattoma) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:314]: I have some doubts about if it is necessary to check if the builds failed. And if they did, then restart them. We could discuss during automation meeting

Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:314]: The builds are started automatically(by cuda-thoth after its completion) , So Is there any chance that they would fail?

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 6]
```diff
@@ -153,6 +155,25 @@ Verify "Notebook Images Are Building" Is Not Shown When No Images Are Building
...
+    Verify Notification Saying Notebook Builds Not Started
+    Clear Dashboard Notifications
+    Wait Until Build Status Is    namespace=redhat-ods-applications    build_name=${last_cuda_build}  expected_status=Complete
+    Remove Values From List    ${images}  CUDA
+    RHODS Notification Drawer Should Contain    message=Notebook images are building
+    Sleep  20s  reason=Wait For Minimal Cuda Build to Start and Notifications
+    Verify Notification Saying Notebook Builds Not Started
+    RHODS Notification Drawer Should Not Contain    message=CUDA
+    ${minimal_cuda_build}=  Search Last Build    namespace=redhat-ods-applications    build_name_includes=minimal
+    Wait Until Build Status Is    namespace=redhat-ods-applications    build_name=${minimal_cuda_build}
```
[the remark this answers is not in the document]
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Updated.

File: tests/Resources/Page/ODH/ODHDashboard/ODHDashboard.robot [hunk 7]
```diff
@@ -607,3 +607,13 @@ RHODS Notification Drawer Should Not Contain
...
+
+Clear Dashboard Notifications
+    [Documentation]     Clears Notifications present in RHODS dashboard
+    Click Element    xpath=//*[contains(@class,'notification-badge')]
+    Sleep  5s  reason=To avoid element not interactable exception
+    ${notification_count}=  Get Element Count    class:odh-dashboard__notification-drawer__item-remove
+    FOR    ${index}    IN RANGE    ${notification_count}
+        Click Element    xpath=//*[contains(@class,"odh-dashboard__notification-drawer__item-remove")]
+    END
+    Click Element    xpath=//*[contains(@class,'notification-badge')]
```
Comment (lugi0) [inline tests/Resources/Page/ODH/ODHDashboard/ODHDashboard.robot:619]: You can use `Close Notification Drawer`

**Review by lugi0 — CHANGES_REQUESTED** [review]

File: tests/Resources/Page/ODH/ODHDashboard/ODHDashboard.robot (hunk 7 above)
Comment (prudhvi-165) [reply tests/Resources/Page/ODH/ODHDashboard/ODHDashboard.robot]: Updated

File: tests/Resources/Page/ODH/ODHDashboard/ODHDashboard.robot [hunk 8]
```diff
@@ -432,9 +432,9 @@ RHODS Notification Drawer Should Contain
     [Documentation]    Verifies RHODS Notifications contains given Message
     [Arguments]     ${message}
     Click Element    xpath=//*[contains(@class,'notification-badge')]
-    Page Should Contain Element
-    ...    xpath=//*[contains(text(),'${message}')]
-    Close Notification Drawer
+    Page Should Contain  text=${message}
+    Click Element    xpath=//*[contains(@class,'notification-badge')]
+
```
[the remark this answers is not in the document]
Comment (prudhvi-165) [reply tests/Resources/Page/ODH/ODHDashboard/ODHDashboard.robot]: Updated

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 5 above)
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Added A Check for Failing Builds

File: tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot [hunk 9]
```diff
@@ -66,6 +66,7 @@ Verify A Default Image Is Provided And Starts Successfully
     @{user_data} =    Get Previously Selected Notebook Image Details
     Launch JupyterHub Spawner From Dashboard
     Spawn Notebook
+    Sleep  7
```
Comment (lugi0) [inline tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot:69]: This should be `Wait for JupyterLab Splash Screen  timeout=30`

Comment (prudhvi-165) [reply tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot]: Updated.

**Review by lugi0 — DISMISSED** [review]
lgtm

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/835/

Result: Passed

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/898/

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 2 above)
Comment (bdattoma) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: I don't see this change

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 5 above)
Comment (bdattoma) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: yes, but it works only if the build fails only once. If after the restart it fails again I think this code will not catch the failure. However, it might be enough for now..what do you think @jgarciao @lugi0 @tarukumar @pablofelix @mwaykole ? I recall we had a discussion on this topic some automation meetings ago

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot [hunk 10]
```diff
@@ -86,3 +86,38 @@ Wait Until All Builds Are Complete
...
+        ${build_name}=  Search Last Build  namespace=${namespace}    build_name_includes=${build}
+        Delete Build    namespace=${namespace}    build_name=${build_name}
+    END
+
+Start Remaining Builds And Wait Until All Builds Are Complete
+    [Documentation]    Starts new build if build fails or is not started , Waits until all builds are complete
+    ${build_configs} =    Create List    11.4.2-cuda-s2i-base-ubi8    11.4.2-cuda-s2i-core-ubi8
+    ...    11.4.2-cuda-s2i-py38-ubi8    11.4.2-cuda-s2i-thoth-ubi8-py38    s2i-minimal-gpu-cuda-11.4.2-notebook
+    ...    s2i-pytorch-gpu-cuda-11.4.2-notebook    s2i-tensorflow-gpu-cuda-11.4.2-notebook
+    ${builds} =    Create List    cuda-s2i-base    cuda-s2i-core    cuda-s2i-py    cuda-s2i-thoth    minimal    pytorch
```
Comment (bdattoma) [inline tests/Resources/Page/OCPDashboard/Builds/Builds.robot:103]: could you take these two lists as arguments instead of defining inside the kw? In this way the kw becomes more generic. 
Then, you could define those lists in the "Variables" section of the test suite

**Review by bdattoma — CHANGES_REQUESTED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 2 above)
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Updated

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot (hunk 10 above)
Comment (prudhvi-165) [reply tests/Resources/Page/OCPDashboard/Builds/Builds.robot]: Updated

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 1 above)
Comment (bdattoma) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: hm I don't think now the test runs tensorflow and pytoch in parallel, because once it starts a build it waits until completion, so it's not parallel. 
It may be complex implement a logic for running in parallel the builds here because we shoudl consider the dependencies..I think it's not worth doing it now. @lugi0 @jgarciao @tarukumar @pablofelix @mwaykole thoughts?

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/928/

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/946

**Review by bdattoma — APPROVED** [review]

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot [hunk 11]
```diff
@@ -86,3 +86,34 @@ Wait Until All Builds Are Complete
...
+        ${build_name}=  Search Last Build  namespace=${namespace}    build_name_includes=${build}
+        Delete Build    namespace=${namespace}    build_name=${build_name}
+    END
+
+Start Remaining Builds And Wait Until All Builds Are Complete
+    [Documentation]    Starts new build if build fails or is not started , Waits until all builds are complete
+    [Arguments]    ${builds}  ${build_configs}
+    ${no_of_builds} =    Get Length    ${builds}
+    FOR    ${ind}    IN RANGE    ${no_of_builds}
+        ${build_name} =    Search Last Build    namespace=redhat-ods-applications
```
Comment (jgarciao) [inline tests/Resources/Page/OCPDashboard/Builds/Builds.robot:103]: Can you please replace redhat-ods-applications for a ${namespace} parameter?

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot [hunk 12]
```diff
@@ -86,3 +86,34 @@ Wait Until All Builds Are Complete
...
+        ...    build_name_includes=${builds}[${ind}]
+        IF    "${build_name}" == ""
+            ${build_name} =    Start New Build    namespace=redhat-ods-applications
+            ...    buildconfig=${build_configs}[${ind}]
+        ELSE
+            ${build_status} =    Get Build Status    namespace=redhat-ods-applications
+            ...    build_search_term=${build_name}
+            IF    "${build_status}" == "Failed"
```
Comment (jgarciao) [inline tests/Resources/Page/OCPDashboard/Builds/Builds.robot:111]: build_status  can be also == "Error" when builds don't work well

File: tests/Tests/500__jupyterhub/special-user-testing.robot [hunk 13]
```diff
@@ -19,7 +19,7 @@ Suite Teardown   End Web Test
 Test Special Usernames
     [Tags]  Smoke
     ...     PLACEHOLDER  #Category tags
-    ...     ODS-257
+    ...     ODS-257  ODS-532
```
Comment (jgarciao) [inline tests/Tests/500__jupyterhub/special-user-testing.robot:22]: This change seems unrelated to this PR

File: tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot [hunk 14]
```diff
@@ -66,6 +67,7 @@ Verify A Default Image Is Provided And Starts Successfully
     @{user_data} =    Get Previously Selected Notebook Image Details
     Launch JupyterHub Spawner From Dashboard
     Spawn Notebook
+    Wait For JupyterLab Splash Screen  timeout=30
```
Comment (jgarciao) [inline tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot:70]: This change seems unrelated to this PR

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 1 above)
Comment (jgarciao) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: I think it's better to keep it simple, even if it takes longer. Also, we'll have pre-built images in the near future

File: tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot (hunk 14 above)
Comment (prudhvi-165) [reply tests/Tests/500__jupyterhub/test-jupyterlab-notebook.robot:70]: Yes, it is related to one of the previous PR which was merged but was failing because of not waiting for jupyterlab splash screen

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 5 above)
Comment (jgarciao) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: Seems to be addressed now with the new Test Teardown rebuilding all failed builds

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot [hunk 15]
```diff
@@ -86,3 +86,34 @@ Wait Until All Builds Are Complete
...
+
+Delete Multiple Builds
+    [Documentation]     Deletes Multiple Builds in the given namespace
+    [Arguments]   @{build_name_list}  ${namespace}
+    FOR    ${build}    IN    @{build_name_list}
+        ${build_name}=  Search Last Build  namespace=${namespace}    build_name_includes=${build}
+        Delete Build    namespace=${namespace}    build_name=${build_name}
+    END
+
+Start Remaining Builds And Wait Until All Builds Are Complete
```
Comment (jgarciao) [inline tests/Resources/Page/OCPDashboard/Builds/Builds.robot:98]: Could be renamed to  `Rebuild Missing Or Failed Builds`

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot (hunk 15 above)
Comment (prudhvi-165) [reply tests/Resources/Page/OCPDashboard/Builds/Builds.robot]: Updated

File: tests/Tests/500__jupyterhub/special-user-testing.robot (hunk 13 above)
Comment (prudhvi-165) [reply tests/Tests/500__jupyterhub/special-user-testing.robot]: Updated

File: tests/Resources/Page/OCPDashboard/Builds/Builds.robot (hunk 12 above)
Comment (prudhvi-165) [reply tests/Resources/Page/OCPDashboard/Builds/Builds.robot]: Updated

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/988/

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 16]
```diff
@@ -188,6 +197,23 @@ Verify Favorite Resource Cards
     Favorite Items Should Be Listed First When Sorted By    ${favorite_ids}    duration
     [Teardown]    Remove Items From favorites    @{favorite_ids}
 
+Verify Notifications Are Shown When Notebook Builds Have Not Started
+    [Documentation]     Verifies that Notifications are shown in RHODS Dashboard when Notebook builds haven't started
+    [Tags]    Sanity
+    ...       ODS-1347
+    ...       Tier1
```
Comment (jgarciao) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:204]: As the test took 48 minutes, could you add this tag? `Execution-Time-Over-30m`

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 16 above)
Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:204]: Updated

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 17]
```diff
@@ -188,6 +197,24 @@ Verify Favorite Resource Cards
...
```
Comment (jgarciao) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:204]: Sorry I didn't think about this yesterday but as  execution is 48 minutes I think it should be in Tier2, not Tier1

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 17 above)
Comment (bdattoma) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:204]: I would say Tier3 where all the long running tests from all the other Tiers are placed

Comment (prudhvi-165) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot]: updated.

**Review by bdattoma — APPROVED** [review]

**Comment by jgarciao** [comment]
@prudhvi-165 Can you solve the merge conflicts?

**Comment by prudhvi-165** [comment]
> @prudhvi-165 Can you solve the merge conflicts?

Done

**Comment by prudhvi-165** [comment]
> @prudhvi-165 pls could you test it once again before we can merge it?

Jenkins Link:https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/1051/

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot [hunk 18]
```diff
@@ -32,10 +33,19 @@ ${RHOSAK_DISPLAYED_APPNAME}             OpenShift Streams for Apache Kafka
...
+@{BUILDS_TO_BE_CHECKED}                 cuda-s2i-base    cuda-s2i-core    cuda-s2i-py    cuda-s2i-thoth
+...                                     minimal    pytorch  tensorflow
 ${openvino_appname}           ovms
 ${openvino_container_name}    OpenVINO
 ${openvino_operator_name}     OpenVINO Toolkit Operator
 
+
 *** Test Cases ***
 Verify That Login Page Is Shown When Reaching The RHODS Page
     [Tags]      Sanity
```
Comment (jgarciao) [inline tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:51]: Tier1 missing

**Review by jgarciao — APPROVED** [review]

File: tests/Tests/400__ods_dashboard/400__ods_dashboard.robot (hunk 18 above)
Comment (jgarciao) [reply tests/Tests/400__ods_dashboard/400__ods_dashboard.robot:51]: To be fixed in another PR (this test is not new from this PR)
